SSA/Projects/CSE/CSE.lean

```diff
@@ -0,0 +1,526 @@
+        apply (Subtype.mk v'.toSnoc)
+        intros V
+        simp [Lets.denote_lete]
+        rw [hv']
+
+      case last => -- new variable, return the CSE'd variable.
+        apply (Subtype.mk vold.toSnoc)
+        intros V
+        simp [Lets.denote_lete]
+        rw [← hv]
+        rw [henew]
+    expr2cache := fun β eneedle =>
+      let homRemap := Ctxt.Hom.remapLast Γ vold
+      let lastVar := (Ctxt.Var.last Γ α)
+      let ⟨eneedle', heneedle'⟩ := ExprRemapVar lets homRemap vold lastVar (by {
+        intros Vstart
+        simp [Ctxt.Hom.remapLast, Ctxt.Valuation.comap]
+        done
+      })  eneedle
+      match s.expr2cache β eneedle' with
+      | .none => .none
+      | .some ⟨e', he'⟩ =>
+        .some ⟨e', by {
+          intros V
+          simp
+          rw [he']
+          rw [heneedle']
+          congr
+          funext ty var
+          cases var using Ctxt.Var.casesOn
+          case e_Γv.h.h.toSnoc v =>
+            simp [Ctxt.Valuation.comap, Ctxt.Hom.remapLast]
+            done
+          case e_Γv.h.h.last =>
+            simp [Ctxt.Valuation.comap, Ctxt.Hom.remapLast]
+            rw [henew]
+            rw [hv]
+            done
+        }⟩
+}
+
+/-- Replace the variables in `as` with new variables that have the same valuation -/
+def State.cseArgList
+ [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty]
+  {Γstart Γ : Ctxt Ty} {lets : Lets Op Γstart Γ} (s : State Op lets)
+  {ts : List Ty}
+  (as : HVector (Ctxt.Var Γ) <| ts) :
+  { as' : HVector (Ctxt.Var Γ) <| ts // ∀ (V : Γstart.Valuation), as.map (lets.denote V).eval = as'.map (lets.denote V).eval  } :=
+  match as with
+  | .nil => ⟨.nil, by
+      simp [HVector.map]
+    ⟩
+  | .cons a as =>
+    let ⟨a', ha'⟩ :=  s.var2var a
+    let ⟨as', has'⟩ := s.cseArgList as
+    ⟨.cons a' as', by
+          intros V
+          simp [HVector.map]
+          constructor
+          apply ha'
+          apply has'
+    ⟩
+
+/-- Default instance for `partial def` to compile. -/
+instance [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty] : Inhabited (
+          {lets : Lets Op Γstart Γ} →
+          State Op lets →
+          {ts : List (Ctxt Ty × Ty)} →
+          (rs : HVector (fun t => Com Op t.1 t.2) ts) →
+          { rs' : HVector (fun t => Com Op t.1 t.2) ts // HVector.denote rs = HVector.denote rs' }) where
+  default := fun _s _ts rs => ⟨rs, rfl⟩
+
+/-- Default instance for `partial def` to compile. -/
+instance [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty] : Inhabited (
+  {lets : Lets Op Γstart Γ} →
+  State Op lets →
+  (com: Com Op Γ α) →
+  { com' : Com Op Γ α // ∀ (V: Ctxt.Valuation Γ), com.denote V = com'.denote V }) where
+  default := fun _s com => ⟨com, by intros V; rfl⟩
+
+
+/- CSE for HVector / Com / Expr. -/
+mutual
+variable [DecidableEq Ty] [DecidableEq Op] [Goedel Ty] [OpSignature Op Ty] [OpDenote Op Ty]
+
+/-- Replace the regions in `rs` with new regions that have the same valuation -/
+unsafe def State.cseRegionArgList
```

Comment:
why do we need this `unsafe`?

Reply:
Lean is unable to determine that the type is inhabited. I'll lift this restriction as I did in DCE: We need to show that the type has an `Inhabited` instance.

Reply:
Resolved, there is no more `unsafe`.